proposals/0014-wave-size-range.md

```diff
@@ -0,0 +1,283 @@
+# Wave Size Range
+
+* Proposal: [0014](0014-wave-size-range.md)
+* Author(s): [Greg Roth](https://github.com/pow2clk)
+* Sponsor: [Greg Roth](https://github.com/pow2clk)
+* Status: **Under Consideration**
+* Planned Version: Shader Model 6.8, validator version 1.8
+
+## Introduction
+
+Shader Model 6.6 included the ability to specify a wave size on a shader entry
+ point in order to indicate that this shader depends on a specific wave size.
+If that size of wave isn't available to the current hardware,
+ shader will fail to load.
+Existing queries allow the developer to determine if this will happen in
+ advance.
+This enables creating custom shaders optimized for the wave size of a
+ particularly interesting platform and loading them when possible.
+
+## Motivation
+
+Shader Model 6.6 provides no mechanism for specifying multiple wave sizes that a
+ single shader might be able to support.
+Consequently, if more than one target wave size is of interest, separate entry
+ points will have to be created for each wave size.
+These will each have to be compiled and shipped separately and, at runtime,
+ the appropriate shaders will have to be selected.
+This increases the size of the shipped product and potentially slows down
+ runtime shader loading.
+
+Ranges of wave sizes are already present in areas of the D3D API.
+The query to determine what wave sizes are supported takes a range and similar
+ information is transmitted from the shader to the runtime, though not the
+ driver.
+
+## Proposed solution
+
+Modifying the parameters of the `WaveSize` attribute that was introduced by
+ Shader Model 6.6 to take additional parameters will allow the shader author to
+ specify fully what the shader supports and what it prefers.
+
+By adding an optional second `WaveSize` parameter, shader authors can provide
+ two values that represent the range of wave sizes that the shader can support.
+This allows the same shader to be used in the event of the availability of a
+ wave size that the shader has optimized for specifically as well as other sizes
+ that are supported, but might not be so optimal.
+
+Some platforms that support a range of wave sizes might overlap with the range
+ specified as supported by the shader in more than one value.
+In these cases, the runtime has a choice of wave sizes that it probably doesn't
```

Comment:
nit: it wouldn't be the runtime choosing but rather the driver.

Reply:
From HLSL's perspective the distinction between runtime and driver isn't really significant. It may also vary based on which runtime/driver we're talking about.

Probably a more important distinction for us to be making is _when_ this can or should happen (i.e. after PSO creation).